README.md
# Alien::MSYS ![linux](https://github.com/PerlAlien/Alien-MSYS/workflows/linux/badge.svg) ![windows-share](https://github.com/PerlAlien/Alien-MSYS/workflows/windows-share/badge.svg)

Tools required for GNU style configure scripts on Windows

# SYNOPSIS

from Perl:

```perl
use Alien::MSYS;
# runs uname from MSYS
my $uname = msys { `uname` };
```

From Prompt/Makefile

```
C:\> perl -MAlien::MSYS -e Alien::MSYS::msys_run uname
```

# DESCRIPTION

MSYS provides minimal shell and POSIX tools on Windows to enable GNU style configure
scripts to run (the type usually generated by `autoconf`). This module aims to
provide an interface for using MSYS on Windows and act as a no-op on Unix like
operating systems which already have that capability.  If you use this module, I
recommend that you list this as a prerequisite only during MSWin32 installs.

When installing, this distribution will look for an existing `MSYS` using the following
methods in this order:

- environment variable `ALIEN_INSTALL_TYPE` or `ALIEN_MSYS_INSTALL_TYPE`

    If set to `share` a system install will not be attempted.  If set to `system`
    then a share install will not be attempted.

- environment variable `PERL_ALIEN_MSYS_BIN`

    If set, this environment variable should be set to the root of `MSYS` (NOT `MinGW`).
    For example, if you have `MinGW` / `MSYS` installed on `D:` you might use this:

    ```
    C:\> set PERL_ALIEN_MSYS_BIN=D:\MinGW\msys\1.0\bin
    ```

    Keep in mind that this environment variable is consulted during both install and at run-time,
    so it is advisable to set this in the System Properties control panel.

- search `PATH` for `mingw-get.exe`

    Second, [Alien::MSYS](https://metacpan.org/pod/Alien::MSYS) searches the `PATH` environment variable for the `mingw-get.exe`
    program, which is a common method for installing `MinGW` and `MSYS`.  From there
    if it can deduce the location of `MSYS` it will use that.

- try `C:\MinGW\msys\1.0\bin`

    Next, [Alien::MSYS](https://metacpan.org/pod/Alien::MSYS) tries the default install location.

- Use desktop shortcut for `MinGW Installer`

    Finally, [Alien::MSYS](https://metacpan.org/pod/Alien::MSYS) will try to find `MSYS` from the desktop shortcut created
    by the GUI installer for `MinGW`.  This method only works if you already have
    [Win32::Shortcut](https://metacpan.org/pod/Win32::Shortcut) installed, as it is an optional dependency.

If `MSYS` cannot be found using any of these methods, then it will download and install
`MSYS` in this distribution's share directory.

# FUNCTIONS

## msys

```perl
# get the uname from MSYS
my $uname = msys { `uname` };

# run with GNU make from MSYS instead of
# dmake from Strawberry Perl
msys { system 'make' };
```

This function takes a single argument, a code reference, and runs it with the correctly
set environment so that calls to the system function or the qx quote like operator will
use MSYS instead of the default environment.

## msys\_run

```perl
# pass command through @ARGV
C:\> perl -MAlien::MSYS -e Alien::MSYS::msys_run uname

# pass command through @_
C:\> perl -MAlien::MSYS -e "Alien::MSYS::msys_run 'make'; Alien::MSYS::msys_run 'make install'"
```

This function runs a command with the MSYS environment.  It gets the command and arguments
either as passed to it, or if none are passed the the command is expected to be in
@ARGV.

If the command fails then it will [exit](https://metacpan.org/pod/perlfunc#exit) with a non-zero error code.  This
is useful, in the second example above if either `make` or `make install` fails, then
the whole command will fail, also `make install` will not be attempted unless `make`
succeeds.

## msys\_path

This function returns the full path to the MSYS bin directory.

# CAVEATS

This [Alien](https://metacpan.org/pod/Alien) is big and slow to install.  I am aware this is an annoying problem.
It is also the only reliable way (that I know of) to install packages from source
that use autotools on Strawberry or Visual C++ Perl.  Here are some things that you
can do to avoid this painful cost:

- Use the system library if possible

    The [Alien::Build](https://metacpan.org/pod/Alien::Build) system is usually smart enough to find the system library
    if it is available.  [Alien::MSYS](https://metacpan.org/pod/Alien::MSYS) is usually only necessary for so called
    `share` installs.

- Pre-install MSYS

    As mentioned above if you preinstall MSYS and set the `PERL_ALIEN_MSYS_BIN`
    environment variable, then you will save yourself some time if you use multiple
    installs of Perl like I do.

- Use another build system

    Some projects will provide a makefile that will work with GNU Make and `cmd.exe`
    that you can build without MSYS.  An example of an Alien that takes advantage of
    this is [Alien::libuv](https://metacpan.org/pod/Alien::libuv).

    Some projects provide both autoconf and CMake.  Although using CMake reliably
    requires [Alien::cmake3](https://metacpan.org/pod/Alien::cmake3) for `share` installs, it is much much lighter than [Alien::MSYS](https://metacpan.org/pod/Alien::MSYS).

    Also obviously you can open a ticket, or make a pull request with the project that you
    are alienizing to support build systems that don't suck as much as autoconf.

- Use MSYS2

    Strawberry Perl is convenient for building XS modules without any dependencies
    or just dependencies on the small number of libraries that come bundled with
    Strawberry Perl.  It is very very painful in my opinion if you depend on libraries
    that are not bundled, which is why this Alien exists.  There is an alternative though.

    MSYS2 / MinGW provides a MSWin32 Perl as part of a Linux / open source like package
    that provides probably all of the libraries that you might want to use as dependencies,
    and if it doesn't you can build much easier than with Strawberry + Alien::MSYS.

    There are some notes here:

    [https://project-renard.github.io/doc/development/meeting-log/posts/2016/05/03/windows-build-with-msys2/](https://project-renard.github.io/doc/development/meeting-log/posts/2016/05/03/windows-build-with-msys2/)

    On using the MSYS2 / MinGW / MSWin32 Perl from the MSYS2 project.

# AUTHOR

Author: Graham Ollis <plicease@cpan.org>

Contributors:

Shawn Laffan (SLAFFAN, shawnlaffan)

bekuno

Zaki Mughal (zmughal)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013-2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.17      2022-09-25 18:08:16 -0600
  - Documentation fixes and improvements (gh#21)
  - Added msys-coreutils-ext and msys-mktemp as these are commonly
    needed tools (gh#23 zmughal++)

0.16      2022-09-22 06:29:15 -0600
  - Documentation fixes and improvements (gh#19)
  - Update tests to work with ALIEN_DOWNLOAD_RULE restrictions (gh#20)

0.15      2022-08-11 09:51:17 -0600
  - Documentation fixes and improvements

0.14      2020-03-08 17:07:11 -0600
  - Force a system install on non-MSWin32 and non-cygwin as a share
    install doesn't make sense on non-Windows. (SLAFFAN++ gh#9, gh#10)

0.12      2020-02-06 14:46:30 -0700
  - Require at least Windows 7.  Windows XP is known to have poor
    native crypto support, which is used my mingw-get and doesn't
    work with sourceforge.  I recommend that you upgrade to Windows
    10 or better.

0.11      2019-12-18 05:46:43 -0700
  - Production release identical to 0.10_07

0.10_07   2019-12-17 19:29:34 -0700
  - Additional diagnostics

0.10_06   2019-12-16 17:07:33 -0700
  - Additional diagnostics

0.10_05   2019-12-15 18:55:16 -0700
  - All detecton logic has been moved from Alien/MSYS.pm into
    the alienfile.  This is how most Alien::Build Aliens work,
    so it makes more sense.
  - Remove mingw-get cache, man pages and documentation which saves
    about 35MB (1/3 of the total) on a share install.

0.10_04   2019-12-15 07:02:23 -0700
  - Bump EUMM requirements

0.10_03   2019-12-11 16:11:24 -0700
  - Prefer curl to HTTP::Tiny when Net::SSLeay and IO::Socket::SSL
    are not installed.
  - Fix tests to not rely on `touch` which is frequently not available
    on Windows.

0.10_02   2017-08-23 14:11:48 -0400
  - Use MSYS2 if found, instead of downloading MSYS

0.10_01   2017-08-18 17:46:48 -0400
  - Use Alien::Build + alienfile for build
  - Switch from Module::Build (MB) to ExtUtils::MakeMaker (EUMM)
  - Remove dependency on File::ShareDir

0.09      2017-07-23 00:04:49 -0400
  - Remove dependency on File::HomeDir

0.08      2017-06-27 05:02:40 -0400
  - Add environment override ALIEN_MSYS_INSTALL_TYPE, ALIEN_INSTALL_TYPE

0.07      2014-05-15 09:17:18 -0400
  - fix reinstall bug (GH#3)

0.06      2014-05-15 04:00:21 -0400
  - Fix build bug where Win32::Shortcut is installed
    but no desktop icon for the MSYS GUI installer (GH#3)

0.05      2014-05-08 08:54:54 -0400
  - documentation tweaks
  - search PATH for mingw-setup.exe as a method
    for discovering the msys root

0.04      2013-12-13 16:39:45 -0500
  - First CPAN release

0.03      2013-12-13 14:36:20 -0500
  - MSYS now downloaded and installed if not found

0.02      2013-12-08 16:08:37 -0500
  - documentation
  - msys_run does an exit(non-zero) on child failure

0.01      2013-12-07 13:37:22 -0500
  - initial version


inc/My/ModuleBuild.pm
package My::ModuleBuild;

use strict;
use warnings;
use base qw( Module::Build );
use File::Path qw( mkpath );
use File::Spec;
use Env qw( @PATH );
use FindBin ();
use File::Temp qw( tempdir );
use Cwd qw( cwd );

sub _fetch_index1
{
  require HTTP::Tiny;  
  my $url = 'http://sourceforge.net/projects/mingw/files/Installer/mingw-get/';
  my $index = HTTP::Tiny->new->get($url);
  
  $index->{status} =~ /^2..$/ || die join(' ', $index->{status}, $index->{reason}, $url);
  
  my $link;
  
  for($index->{content} =~ m{"/(projects/mingw/files/Installer/mingw-get/mingw-get-.*?-(\d\d\d\d\d\d\d\d)-(\d+))/"})
  {
    if(!defined $link || ($link->{date} <= $2 && $link->{num} < $3))
    {
      $link = {
        url  => "http://sourceforge.net/$1",
        date => $2,
        num  => $3,
      };
    }
  }

  die "couldn't find mingw-get in index" unless $link;
  $link->{url};
}

sub _fetch_index2
{
  my(undef, $url) = @_;
  require HTTP::Tiny;  
  my $index = HTTP::Tiny->new->get($url);
  
  $index->{status} =~ /^2..$/ || die join(' ', $index->{status}, $index->{reason}, $url);

  die "couldn't find mingw-get in download index"
    unless $index->{content} =~ m{"(https?://.*/(mingw-get-.*?-bin.zip)/download)"};
    
  ($1, $2);
}

sub _fetch_zip
{
  my(undef, $url) = @_;
  my $download = HTTP::Tiny->new->get($url);
  $download->{status} =~ /^2..$/ || die join(' ', $download->{status}, $download->{reason}, $url);
  $download->{content};
}

sub ACTION_build
{
  my $self = shift;

  return $self->SUPER::ACTION_build(@_)
    if $^O ne 'MSWin32' || do {
      require lib;
      lib->import('lib');
      require Alien::MSYS;
      # TODO: this means on a re-install we have to re-download
      # if it has been a long time since the last version, this may
      # not be a bad thing.  If we end up having lots of revisions
      # this could be highly annoying.
      do { no warnings 'redefine'; *Alien::MSYS::_my_dist_dir = sub {} };
      defined Alien::MSYS::msys_path();
    };

  foreach my $try ($ENV{PERL_ALIEN_MSYS_BIN}, 'C:/MinGW/msys/1.0/bin')
  {
    my $sh_path = File::Spec->catfile($try, 'sh.exe');
    return $self->SUPER::ACTION_build(@_) if -x $sh_path;
  }

  my($url, $zipname) = __PACKAGE__->_fetch_index2(__PACKAGE__->_fetch_index1);
  my $zipcontent = __PACKAGE__->_fetch_zip($url);

  require Archive::Zip;
  
  my $dir = File::Spec->catdir($FindBin::Bin, qw( share ));
  mkpath($dir, 1, 0755);
  
  my $save = cwd();
  chdir $dir;
  
  my $zip = Archive::Zip->new;
  $zip->read(do {
    my $fn = File::Spec->catdir(tempdir(CLEANUP => 0), $zipname);
    open my $fh, '>', $fn;
    binmode $fh;
    print $fh $zipcontent;
    close $fh;
    print "fn = $fn\n";
    $fn;
  });
  $zip->extractTree();
  
  my $dh;
  opendir($dh, File::Spec->catdir($dir, qw( libexec mingw-get )));
  foreach my $file (readdir $dh)
  {
    next unless $file =~ /\.dll$/;
    eval { chmod 0755, $file };
  }
  closedir $dh;
  
  chdir $save;

  push @PATH, File::Spec->catdir($dir, qw( bin ));
  system 'mingw-get', 'install', 'msys';
  
  # A lot of tools also need m4
  system 'mingw-get', 'install', 'msys-m4';
  
  # A number of autotools (autoconf, automake) are
  # implemented in Perl, but when you build them
  # with Perl on windows using MSYS, they use MSYS
  # paths, like /c/ instead of c:/, so yes.  we
  # also need MSYS Perl.  *sigh*
  system 'mingw-get', 'install', 'msys-perl';

  $self->SUPER::ACTION_build(@_);
}

1;


lib/Alien/MSYS.pm
package Alien::MSYS;

use strict;
use warnings;
use Path::Tiny ();
use Env qw( @PATH );
use base qw( Alien::Base Exporter );

our @EXPORT    = qw( msys msys_run );
our @EXPORT_OK = qw( msys msys_run msys_path );

# ABSTRACT: Tools required for GNU style configure scripts on Windows
# VERSION

=head1 SYNOPSIS

from Perl:

 use Alien::MSYS;
 # runs uname from MSYS
 my $uname = msys { `uname` };

From Prompt/Makefile

 C:\> perl -MAlien::MSYS -e Alien::MSYS::msys_run uname

=head1 DESCRIPTION

MSYS provides minimal shell and POSIX tools on Windows to enable GNU style configure
scripts to run (the type usually generated by C<autoconf>). This module aims to
provide an interface for using MSYS on Windows and act as a no-op on Unix like
operating systems which already have that capability.  If you use this module, I
recommend that you list this as a prerequisite only during MSWin32 installs.

When installing, this distribution will look for an existing C<MSYS> using the following
methods in this order:

=over 4

=item environment variable C<ALIEN_INSTALL_TYPE> or C<ALIEN_MSYS_INSTALL_TYPE>

If set to C<share> a system install will not be attempted.  If set to C<system>
then a share install will not be attempted.

=item environment variable C<PERL_ALIEN_MSYS_BIN>

If set, this environment variable should be set to the root of C<MSYS> (NOT C<MinGW>).
For example, if you have C<MinGW> / C<MSYS> installed on C<D:> you might use this:

 C:\> set PERL_ALIEN_MSYS_BIN=D:\MinGW\msys\1.0\bin

Keep in mind that this environment variable is consulted during both install and at run-time,
so it is advisable to set this in the System Properties control panel.

=item search C<PATH> for C<mingw-get.exe>

Second, L<Alien::MSYS> searches the C<PATH> environment variable for the C<mingw-get.exe>
program, which is a common method for installing C<MinGW> and C<MSYS>.  From there
if it can deduce the location of C<MSYS> it will use that.

=item try C<C:\MinGW\msys\1.0\bin>

Next, L<Alien::MSYS> tries the default install location.

=item Use desktop shortcut for C<MinGW Installer>

Finally, L<Alien::MSYS> will try to find C<MSYS> from the desktop shortcut created
by the GUI installer for C<MinGW>.  This method only works if you already have
L<Win32::Shortcut> installed, as it is an optional dependency.

=back

If C<MSYS> cannot be found using any of these methods, then it will download and install
C<MSYS> in this distribution's share directory.

=head1 FUNCTIONS

=head2 msys

 # get the uname from MSYS
 my $uname = msys { `uname` };
 
 # run with GNU make from MSYS instead of
 # dmake from Strawberry Perl
 msys { system 'make' };

This function takes a single argument, a code reference, and runs it with the correctly
set environment so that calls to the system function or the qx quote like operator will
use MSYS instead of the default environment.

=cut

sub msys (&)
{
  local $ENV{PATH} = $ENV{PATH};
  if($^O eq 'MSWin32')
  {
    unshift @PATH, msys_path();
  }
  $_[0]->();
}

=head2 msys_run

 # pass command through @ARGV
 C:\> perl -MAlien::MSYS -e Alien::MSYS::msys_run uname
 
 # pass command through @_
 C:\> perl -MAlien::MSYS -e "Alien::MSYS::msys_run 'make'; Alien::MSYS::msys_run 'make install'"

This function runs a command with the MSYS environment.  It gets the command and arguments
either as passed to it, or if none are passed the the command is expected to be in
@ARGV.

If the command fails then it will L<exit|perlfunc#exit> with a non-zero error code.  This
is useful, in the second example above if either C<make> or C<make install> fails, then
the whole command will fail, also C<make install> will not be attempted unless C<make>
succeeds.

=cut

sub msys_run
{
  my $cmd = \@_;
  $cmd = \@ARGV unless @$cmd > 0;
  msys { system @$cmd };
  # child did an exit(0), or "success"
  return if $? == 0;
  # child wasn't able to exit (-1) or died with signal
  exit 2 if $? == -1 || $? & 127;
  # child exited with non zero
  exit $?;
}

=head2 msys_path

This function returns the full path to the MSYS bin directory.

=cut

sub msys_path ()
{
  my $class = 'Alien::MSYS';

  return if $^O ne 'MSWin32';

  if($class->install_type('share'))
  {
    return Path::Tiny->new($class->dist_dir)->child('msys/1.0/bin')->canonpath;
  }
  else
  {
    return Path::Tiny->new($class->runtime_prop->{my_bin})->canonpath;
  }
}

1;

=head1 CAVEATS

This L<Alien> is big and slow to install.  I am aware this is an annoying problem.
It is also the only reliable way (that I know of) to install packages from source
that use autotools on Strawberry or Visual C++ Perl.  Here are some things that you
can do to avoid this painful cost:

=over 4

=item Use the system library if possible

The L<Alien::Build> system is usually smart enough to find the system library
if it is available.  L<Alien::MSYS> is usually only necessary for so called
C<share> installs.

=item Pre-install MSYS

As mentioned above if you preinstall MSYS and set the C<PERL_ALIEN_MSYS_BIN>
environment variable, then you will save yourself some time if you use multiple
installs of Perl like I do.

=item Use another build system

Some projects will provide a makefile that will work with GNU Make and C<cmd.exe>
that you can build without MSYS.  An example of an Alien that takes advantage of
this is L<Alien::libuv>.

Some projects provide both autoconf and CMake.  Although using CMake reliably
requires L<Alien::cmake3> for C<share> installs, it is much much lighter than L<Alien::MSYS>.

Also obviously you can open a ticket, or make a pull request with the project that you
are alienizing to support build systems that don't suck as much as autoconf.

=item Use MSYS2

Strawberry Perl is convenient for building XS modules without any dependencies
or just dependencies on the small number of libraries that come bundled with
Strawberry Perl.  It is very very painful in my opinion if you depend on libraries
that are not bundled, which is why this Alien exists.  There is an alternative though.

MSYS2 / MinGW provides a MSWin32 Perl as part of a Linux / open source like package
that provides probably all of the libraries that you might want to use as dependencies,
and if it doesn't you can build much easier than with Strawberry + Alien::MSYS.

There are some notes here:

L<https://project-renard.github.io/doc/development/meeting-log/posts/2016/05/03/windows-build-with-msys2/>

On using the MSYS2 / MinGW / MSWin32 Perl from the MSYS2 project.

=back

=cut
